sklearn/datasets/_base.py

```diff
@@ -1121,6 +1121,19 @@ def load_boston(*, return_X_y=False):
 
     Read more in the :ref:`User Guide <boston_dataset>`.
 
+    .. warning::
+       The Boston housing prices dataset has an ethical problem: as
+       investigated in [1]_, the authors of this dataset engineered a
+       non-invertible variable "B" assuming that racial self-segregation had a
+       positive impact on house prices [2]_. Furthermore the goal of the
+       research that led to the creation of this dataset was to study the
+       impact of air quality but it did not give adequate demonstration of the
+       validity of this assumption.
+
+       The scikit-learn maintainers therefore strongly discourage the use of
+       this dataset unless the purpose of the code is to study and educate
+       about ethical issues in data science and machine learning.
+
```

Comment:
Maybe we should also propose alternative datasets to help "operationalize" this recommendation. For instance:

> Alternative datasets include the California housing dataset (`fetch_california_housing`) or the Ames housing > dataset that can be fetched with `fetch_openml(name="house_prices", as_frame=True)`.

And include links to the main references.